### PR TITLE
Always use soft-deletes in InternalEngine

### PR DESCRIPTION
**Pull request by dnhatn**
Peer recoveries become faster and use less storage (i.e., no more extra translog) with soft-deletes. Soft-deletes has been enabled by default since 7.0. We should make it mandatory in 8.0, so we can simplify the logic in the engine, recoveries, and other components.

With this change, InternalEngine will always use soft-deletes regardless of the soft_deletes settings.

This PR will go to 8.0 only. I will open another PR to deprecate the soft_deletes setting (and adapt this change in other components).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -2200,11 +2177,7 @@ private IndexWriter createWriter() throws IOException {
 
     // pkg-private for testing
     IndexWriter createWriter(Directory directory, IndexWriterConfig iwc) throws IOException {
-        if (Assertions.ENABLED) {
-            return new AssertingIndexWriter(directory, iwc);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: why not keep `AssertingIndexWriter`? Perhaps still useful to check that updateDocument(s) / deleteDocuments / tryDeleteDocument is never called?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -2441,9 +2412,7 @@ protected void commitIndexWriter(final IndexWriter writer, final Translog transl
                 commitData.put(SequenceNumbers.MAX_SEQ_NO, Long.toString(localCheckpointTracker.getMaxSeqNo()));
                 commitData.put(MAX_UNSAFE_AUTO_ID_TIMESTAMP_COMMIT_ID, Long.toString(maxUnsafeAutoIdTimestamp.get()));
                 commitData.put(HISTORY_UUID_KEY, historyUUID);
-                if (softDeleteEnabled) {
-                    commitData.put(Engine.MIN_RETAINED_SEQNO, Long.toString(softDeletesPolicy.getMinRetainedSeqNo()));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: how do we deal with the possible absence of this information on older commits?

**Review by ywelsch — COMMENTED** [review]
Looking good. Thanks @dnhatn!

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: ok, restored in https://github.com/elastic/elasticsearch/pull/50415/commits/60139407229a51c670a67a4f594e4b239b2d3e04.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I pushed https://github.com/elastic/elasticsearch/pull/50415/commits/aa7e6ef271c602f5d19c31148e6e0a55088203b1 to set min_retained_seq_no to the max_seq_no of the safe commit. I also enabled the history assertion for non-soft-deletes indices.

**Review by ywelsch — COMMENTED** [review]
I also wonder how the changes here impacts `ReadOnlyEngine` which has a special wrapper for soft-deleted indices.

**Comment by dnhatn** [comment]
> I also wonder how the changes here impacts ReadOnlyEngine which has a special wrapper for soft-deleted indices.

Good call. Wrapping a hard-deletes reader in a soft-deletes reader is safe and trivial. We should always do it. I think I need to go through the usage of the method `isSoftDeleteEnabled()` more.

**Review by ywelsch — APPROVED** [review]
Can you follow this up with a test that checks migration of frozen indices (i.e. full-cluster-upgrade test) from soft-deletes disabled to enabled? Just to be sure that `ReadOnlyEngine` works correctly after upgrading from non-soft-deleted indices.

**Comment by dnhatn** [comment]
> Can you follow this up with a test that checks migration of frozen indices (i.e. full-cluster-upgrade test) from soft-deletes disabled to enabled? Just to be sure that ReadOnlyEngine works correctly after upgrading from non-soft-deleted indices.

Sure, will do. Thank you for reviewing :).

**Pull request merged**